snuba/query/matcher.py

```diff
@@ -0,0 +1,201 @@
+    """
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult()
+
+
+@dataclass(frozen=True)
+class String(Pattern[Optional[str]]):
+    """
+    Matches one specific string (or None).
+    """
+
+    value: Optional[str]
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[Expression]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each a Pattern can be
+    provided. If one of these Patters is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[Optional[str]]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.column_name, node.column_name),
+            (self.table_name, node.table_name),
+        ):
+            if pattern:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Function(Pattern[Expression]):
+    """
+    Matches a Function in the AST expression.
+    It works like the Column Pattern. if alias, function_name and function_parameters
+    are provided, they have to match, otherwise they are ignored.
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    function_name: Optional[Pattern[Optional[str]]] = None
+    function_parameters: Optional[Tuple[Pattern[Expression], ...]] = None
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        if not isinstance(node, FunctionCall):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.function_name, node.function_name),
+        ):
+            if pattern:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        if self.function_parameters:
+            if len(self.function_parameters) != len(node.parameters):
+                return None
+            for index, param_pattern in enumerate(self.function_parameters):
+                p_result = param_pattern.match(node.parameters[index])
+                if not p_result:
+                    return None
+                else:
+                    result = result.merge(p_result)
+
+        return result
+
+
+# TODO: Add more Patterns when needed.
```

Comment:
I only added the minimum set of useful Pattern subclasses. I. can see we will add quite a few more here.